crates/wasi/src/preview2/pipe.rs

```diff
@@ -7,227 +7,233 @@
-/// let mut table = Table::new();
-/// let stdout = WritePipe::new_in_memory();
-/// let mut ctx = WasiCtxBuilder::new().set_stdout(stdout.clone()).build(&mut table).unwrap();
-/// // use ctx and table in an instance, then make sure it is dropped:
-/// drop(ctx);
-/// drop(table);
-/// let contents: Vec<u8> = stdout.try_into_inner().expect("sole remaining reference to WritePipe").into_inner();
-/// println!("contents of stdout: {:?}", contents);
-/// ```
 #[derive(Debug)]
-pub struct WritePipe<W: Write> {
-    writer: Arc<RwLock<W>>,
+pub struct MemoryInputPipe {
+    buffer: std::io::Cursor<Vec<u8>>,
 }
 
-impl<W: Write> Clone for WritePipe<W> {
-    fn clone(&self) -> Self {
+impl MemoryInputPipe {
+    pub fn new(bytes: impl AsRef<[u8]>) -> Self {
         Self {
-            writer: self.writer.clone(),
+            buffer: std::io::Cursor::new(Vec::from(bytes.as_ref())),
         }
     }
 }
 
-impl<W: Write> WritePipe<W> {
-    /// Create a new pipe from a `Write` type.
-    ///
-    /// All `Handle` write operations delegate to writing to this underlying writer.
-    pub fn new(w: W) -> Self {
-        Self::from_shared(Arc::new(RwLock::new(w)))
-    }
-
-    /// Create a new pipe from a shareable `Write` type.
-    ///
-    /// All `Handle` write operations delegate to writing to this underlying writer.
-    pub fn from_shared(writer: Arc<RwLock<W>>) -> Self {
-        Self { writer }
-    }
-
-    /// Try to convert this `WritePipe<W>` back to the underlying `W` type.
-    ///
-    /// This will fail with `Err(self)` if multiple references to the underlying `W` exist.
-    pub fn try_into_inner(mut self) -> Result<W, Self> {
-        match Arc::try_unwrap(self.writer) {
-            Ok(rc) => Ok(RwLock::into_inner(rc).unwrap()),
-            Err(writer) => {
-                self.writer = writer;
-                Err(self)
-            }
+#[async_trait::async_trait]
+impl HostInputStream for MemoryInputPipe {
+    fn read(&mut self, dest: &mut [u8]) -> Result<(u64, StreamState), Error> {
+        let nbytes = std::io::Read::read(&mut self.buffer, dest)?;
+        let state = if self.buffer.get_ref().len() as u64 == self.buffer.position() {
+            StreamState::Closed
+        } else {
+            StreamState::Open
+        };
+        Ok((nbytes as u64, state))
+    }
+    async fn ready(&mut self) -> Result<(), Error> {
+        if self.buffer.get_ref().len() as u64 > self.buffer.position() {
+            Ok(())
+        } else {
+            futures_util::future::pending().await
         }
     }
-
-    fn borrow(&self) -> std::sync::RwLockWriteGuard<W> {
-        RwLock::write(&self.writer).unwrap()
-    }
 }
 
-impl WritePipe<io::Cursor<Vec<u8>>> {
-    /// Create a new writable virtual pipe backed by a `Vec<u8>` buffer.
-    pub fn new_in_memory() -> Self {
-        Self::new(io::Cursor::new(vec![]))
-    }
+#[derive(Debug, Clone)]
+pub struct MemoryOutputPipe {
+    buffer: std::sync::Arc<std::sync::Mutex<Vec<u8>>>,
 }
```

Comment:
Yeah, we'd have to do some challenging stuff to destroy the ctx, which is harder because the methods need both the ctx and table via WasiView, so we decided against changing this now.

We really only expect this impl to be used for tests.